docs/proposals/apiserver_watch.md

```diff
@@ -0,0 +1,132 @@
+
+
+## Proposal details
+
+We would like the cache to be __per-resource-type__ and __optional__. Thanks to
+it we will be able to:
+  - have different cache sizes for different resources (e.g. bigger cache
+    [= longer history] for pods, which can significantly affect performance)
+  - avoid any overhead for objects that are watched very rarely (e.g. events
+    are almost not watched at all, but there are a lot of them)
+  - filter the cache for each watcher more effectively
+
+If we decide to support watches spanning different resources in the future and
+we have an efficient indexing mechanisms, it should be relatively simple to unify
+the cache to be common for all the resources.
+
+Since we want the watch in apiserver to be optional for different resource
+types, this needs to be self-contained and hidden behind a well defined API.
+It seems that this API can be very simple - it should just expose a method
+for initializing the cache and another one for registering (and starting) a
+new watcher with a given filtering function (matcher).
+
+The rest of this section describes the concrete steps that need to be done
+to implement the proposal.
+
+1. Since we are initially targeting to have per-resource-type watch in
+apiserver, this should probably be incorporated into generic etcd registry:
+"pkg/registry/generic/etcd". Thus we should migrate all the resources to
+use it (it seems it's not the case now e.g. for services).
+
+2. We should define an API for the watch in apiserver (this will be purely
+internal api just within apiserver, so it should be possible to change it
+relatively easy in the future). Also, we should provide the initial
+implementation that will be just redirecting all the requests to etcd, similarly
+to what we are doing today. This simple implementation should be used to enable
+easy transition to the new mechanism later (api unification).
+
+3. The target implementation internally will have a single watch to etcd open,
+responsible for watching all the changes of objects of a given resource type.
+However, this watch can potentially expire at any time and reconnecting can
+return "too old resource version". In that case relisting is necessary.
+To avoid LIST requests coming from all the watcher at that time, we will
+introduce an additional etcd event type:
+[EtcdResync](../../pkg/tools/etcd_helper_watch.go#L36)
+
+  Whenever reslisting will be done to refresh the internal watch to etcd,
+  EtcdResync event will be send to all the watchers. It will contain the
+  full list of all the objects the watcher is interested in (appropriately
+  filtered) as the parameter of this watch event.
+  Thus, we need to create the EtcdResync event, extend watch.Interface and
+  its implementations to support it and handle those events appropriately
+  in places like
+  [Reflector](../../pkg/client/cache/reflector.go)
+
+4. Create a FIFO cache with a given capacity. This should be as simple as
```

Comment:
Why not store the etcd event exactly (to provide the new + old value)?  If this component is providing a unified event stream for a given key, we don't have to change the higher level code that reads deserializes objects from etcd events.

Do we truly need to include the list functionality into watch?  When we discussed it before, we identified that clients need to know both when a resync happens *and* when the resync is done.  I tend to like the current code where watch and list are separate.